File: westcoast/src/main/java/frc/robot/Dinky.java
package frc.robot;

import edu.wpi.first.wpilibj2.command.button.*;
import com.ctre.phoenix.motorcontrol.can.*;
import edu.wpi.first.wpilibj2.command.*;
import edu.wpi.first.wpilibj.drive.*;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import com.kauailabs.navx.frc.AHRS;
import edu.wpi.first.wpilibj.*;
import frc.subsystems.vision.*;
import frc.subsystems.vision.targeting.*;
import frc.subsystems.drive.motorfactory.*;
import frc.maps.*;
import frc.subsystems.drive.*;
import frc.commands.ballhandling.ShooterLogic;
import frc.commands.motors.*;

public class Dinky extends TimedRobot {

  //Robot Map
  private final RobotMap map = new DinkyMap();

  // Drivetrain

  // Subsystem Motors
  private final WPI_VictorSPX intake = new WPI_VictorSPX(7);
  private final WPI_TalonSRX rightShooter = new WPI_TalonSRX(8);
  private final WPI_TalonSRX index = new WPI_TalonSRX(9);
  private final WPI_TalonSRX leftShooter = new WPI_TalonSRX(5);
  private final VictorSP mag = new VictorSP(12);
  // private final WPI_TalonSRX leftShooter = new WPI_TalonSRX(8);
  // private final WPI_TalonSRX rightShooter = new WPI_TalonSRX(9);

  // Sensors
  private final AHRS ahrs = new AHRS(SPI.Port.kMXP);
  private final VisionRunner vision = new RioVisionRunner();

  // Inputs
  private final Joystick joystick = new Joystick(0);
  private final XboxController xboxController = new XboxController(0);
  private final JoystickButton lB = new JoystickButton(xboxController, 5);
  private final JoystickButton lT = new JoystickButton(xboxController, 7);
  private final JoystickButton rB = new JoystickButton(xboxController, 6);
  private final JoystickButton rT = new JoystickButton(xboxController, 8);
  
  // Commands
  // private final GyroDrive teleop = new GyroDrive(drive, joystick, ahrs);
  // private final Command setIntake = new SetMotor(intake, -0.4);

  // Autonomous Commands

  @Override
  public void robotInit() {
    ahrs.reset();
    vision.init();
  }

  @Override
  public void robotPeriodic() {
    SmartDashboard.putNumber("alignX", vision.getAlignX());
  }

  @Override
  public void teleopInit() {
    final Command intakedex = new SetTwoMotors(intake, index, 0.5, true);
    final Command ingazine = new SetTwoMotors(index, mag, 0.4);
    final Command setShooter = new ShooterLogic(rightShooter, leftShooter);
    final Command constFeedShot = new SetMotor(mag, 0.3);
    final Command shoot = new ShooterLogic(rightShooter, leftShooter);

    lB.whileHeld(ingazine);
    lT.whileHeld(intakedex);
    rT.whileHeld((setShooter.raceWith(new WaitCommand(1))).andThen(shoot.alongWith(constFeedShot)));
  }

  @Override
  public void teleopPeriodic() {
    CommandScheduler.getInstance().run();
    SmartDashboard.putNumber("Amps", rightShooter.getSupplyCurrent());
  }
}

File: westcoast/src/main/kotlin/frc/robot/Dinky.kt
package frc.robot

import edu.wpi.first.wpilibj.Joystick
import edu.wpi.first.wpilibj.TimedRobot
import edu.wpi.first.wpilibj.drive.DifferentialDrive
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.CommandScheduler
import edu.wpi.first.wpilibj2.command.button.JoystickButton
import frc.commands.ballhandling.EverythingOut
import frc.commands.ballhandling.IntakeAndIndex
import frc.commands.ballhandling.IntakeOut
import frc.commands.ballhandling.Shooting
import frc.robot.maps.DinkyMap
import frc.robot.maps.RobotMap
import frc.subsystems.BallHandlingSubsystem
import frc.subsystems.drive.TeleopDrive
import frc.subsystems.drive.motorfactory.MotorFactory
import frc.subsystems.drive.motorfactory.MotorFactoryHybrid
import frc.subsystems.RioVisionRunner
import frc.subsystems.VisionRunner

class Dinky : TimedRobot() {
  // Robot Map
  private val map: RobotMap = DinkyMap()

  // Drivetrain
  var hybridFactory: MotorFactory = MotorFactoryHybrid()
  private val leftMain = hybridFactory.create(2, 1, 3)
  private val rightMain = hybridFactory.create(5, 4, 6)
  private val drive = DifferentialDrive(leftMain, rightMain)

  // Subsystems
  private val ballHandler = BallHandlingSubsystem(map)

  // Vision
  private val vision: VisionRunner = RioVisionRunner()

  // Inputs
  private val joystick = Joystick(0)
  private val topTrigger = JoystickButton(joystick, 1)
  private val bottomTrigger = JoystickButton(joystick, 6)

  // private final JoystickButton deepTrigger = new JoystickButton(joystick, 15);
  private val buttonA = JoystickButton(joystick, 3)

  // private final JoystickButton buttonB = new JoystickButton(joystick, 4);
  private val buttonC = JoystickButton(joystick, 5)
  private val buttonD = JoystickButton(joystick, 7)

  // Commands
  private val teleop = TeleopDrive(map, drive, joystick)
  override fun robotInit() {
    drive.isSafetyEnabled = false
    vision.init()
  }

  override fun robotPeriodic() {
    vision.periodic()
  }

  override fun teleopInit() {
    teleop.cancel()
    CommandScheduler.getInstance().registerSubsystem(ballHandler)
    val ballIn: Command = IntakeAndIndex(ballHandler)
    val ballOut: Command = IntakeOut(ballHandler)
    val spitOut: Command = EverythingOut(ballHandler)
    val shoot: Command = Shooting(map, drive, vision, ballHandler)
    topTrigger.whenPressed(ballIn)
    bottomTrigger.whileHeld(ballOut)
    buttonA.cancelWhenPressed(ballIn)
    buttonC.toggleWhenPressed(shoot)
    buttonD.whileHeld(spitOut)
    teleop.schedule()
  }

  override fun teleopPeriodic() {
    CommandScheduler.getInstance().run()
  }
}